File: data_structures_and_algorithms/Code_401/class-09/solutions/interview-02/typescript/src/palindrome.ts
import { LinkedList } from "./LinkedList";
import { reverse } from "./reverse";

export function isPalindrome<T>(ll: LinkedList<T>): boolean {
  // Palindrome is pretty straightforward. We could do back and front indexes,
  // or we can reverse the list and compare each pair of node items!
  let ll1 = ll.head;
  let ll2 = reverse(ll).head;

  // This should look like a very normal ll traversal, but with two lists at
  // the same time.
  while (ll1 && ll2) {
    // Comparison is strict equality between items
    if (ll1.item !== ll2.item) {
      // The first pair of items that aren't equal, we know it's not a pal
      return false;
    }
    ll1 = ll1.next;
    ll2 = ll2.next;
  }

  // All pairs of forward and reverse matched, so it's a palindrome!
  return true;
}

File: generic-code-challenges/list-palindrome/palindrome.py
def palindrome(ll) -> bool:
    """Return whether or not a linked list is a palindrome."""
    backward = LinkedList()
    curr = ll.head
    while curr:
        backward.push(curr.val)
        curr = curr.next

    one = backward.head
    two = ll.head
    while one:
        if one.val != two.val:
            return False
        one = one.next
        two = two.next

    return True
